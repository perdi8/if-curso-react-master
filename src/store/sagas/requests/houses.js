import axios from 'axios';

export function requestGetHouses() {
	return axios({
		method: 'get',
		url: 'https://api-casas.kevinmamaqi.com/api-casas',
	});
}
