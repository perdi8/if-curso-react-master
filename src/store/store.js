import {configureStore, combineReducers} from '@reduxjs/toolkit';
import counterReducer from './toolkit/counter.slice';
import housesReducer from './toolkit/houses/houses.slice';
import {createLogger} from 'redux-logger';
import createSagaMiddleware from '@redux-saga/core';
import {watcherSaga} from './sagas/rootSaga';

const sagaMiddleware = createSagaMiddleware();

const logger = createLogger({
	collapsed: true,
	diff: true,
});

export const combinedReducers = combineReducers({
	counter: counterReducer,
	houses: housesReducer,
});

const store = configureStore({
	reducer: combinedReducers,
	middleware: (getDefaultMiddleware) =>
		getDefaultMiddleware().concat(logger).concat(sagaMiddleware),
});

sagaMiddleware.run(watcherSaga);

export default store;
