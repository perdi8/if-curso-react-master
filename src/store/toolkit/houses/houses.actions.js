import {createAsyncThunk} from '@reduxjs/toolkit';
import axios from 'axios';

export const fetchHouses = createAsyncThunk('houses/fetchHouses', async (params, thunkAPI) => {
	try {
		const response = await axios({
			method: 'get',
			url: 'https://api-casas.kevinmamaqi.com/api-casas',
		});
		return response.data;
	} catch (error) {
		console.error(error);
	}
});
