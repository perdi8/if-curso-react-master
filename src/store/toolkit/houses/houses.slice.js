import {createSlice} from '@reduxjs/toolkit';

// 1. TYPOS DE ACCIONES para definir e identificar
// 2. DATOS DE LAS ACCIONS (parametros)
// 3. REDUCERS cambios que hacemos en el estado

export const HOUSES_STATUS = {
	INITIAL: 'INITIAL',
	PENDING: 'PENDING',
	FULFILLED: 'FULFILLED',
	FAILURE: 'FAILURE',
};

export const housesSlice = createSlice({
	name: 'houses',
	initialState: {
		status: HOUSES_STATUS.INITIAL,
		houses: [],
		selectedLocation: undefined,
		housesToShow: 9,
	},
	reducers: {
		getHouses() {},
		setHouses(state, action) {
			state.selectedLocation = [action.payload[0].lat, action.payload[0].long];
			state.houses = action.payload;
			state.status = HOUSES_STATUS.FULFILLED;
		},
		setSelectedLocation: (state, action) => {
			state.selectedLocation = action.payload;
		},
		setHousesToShow: (state) => {
			state.housesToShow = state.housesToShow + 9;
		},
	},
	// extraReducers: (builder) => {
	// 	builder.addCase(fetchHouses.pending, (state, action) => {
	// 		state.status = HOUSES_STATUS.PENDING;
	// 	});
	// 	builder.addCase(fetchHouses.fulfilled, (state, action) => {
	// 		state.selectedLocation = [action.payload[0].lat, action.payload[0].long];
	// 		state.houses = action.payload;
	// 		state.status = HOUSES_STATUS.FULFILLED;
	// 	});
	// 	builder.addCase(fetchHouses.rejected, (state, action) => {
	// 		state.status = HOUSES_STATUS.ERROR;
	// 		console.log('he recibido respuesta negativa');
	// 	});
	// },
});

// Action creators are generated for each case reducer function
export const {getHouses, setHouses, setSelectedLocation, setHousesToShow} = housesSlice.actions;

export default housesSlice.reducer;
