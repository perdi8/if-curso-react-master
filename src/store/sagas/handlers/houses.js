import {call, put} from 'redux-saga/effects';
import {setHouses} from 'store/toolkit/houses/houses.slice';
import {requestGetHouses} from '../requests/houses';

export function* handleGetHouses(action) {
	try {
		const response = yield call(requestGetHouses);
		yield put({type: setHouses.type, payload: response.data});
	} catch (error) {
		console.error(error);
	}
}
