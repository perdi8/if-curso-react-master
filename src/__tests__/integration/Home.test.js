import React from 'react';
import Home from 'pages/Home/Home';
import {fireEvent, render, waitFor} from '@testing-library/react';
import {Provider} from 'react-redux';
import {combinedReducers} from 'store/store';
import {configureStore} from '@reduxjs/toolkit';
import {HOUSES_STATUS} from 'store/toolkit/houses/houses.slice';
import renderer from 'react-test-renderer';

let store;
beforeEach(() => {
	store = configureStore({reducer: combinedReducers});
});

describe('Home page:', () => {
	it('has INITIAL as the status before it renders for the first time', () => {
		expect(store.getState().houses.status).toBe(HOUSES_STATUS.INITIAL);
	});

	it('has PENDING as the status when the component mounts', () => {
		render(
			<Provider store={store}>
				<Home />
			</Provider>
		);
		expect(store.getState().houses.status).toBe(HOUSES_STATUS.PENDING);
	});

	it('has FULFILLED when the component finishes the request and renders the content', async () => {
		const {getAllByText} = render(
			<Provider store={store}>
				<Home />
			</Provider>
		);
		await waitFor(() => {
			expect(getAllByText('Localizar').length).toEqual(3);
		});
		expect(store.getState().houses.status).toBe(HOUSES_STATUS.FULFILLED);
	});

	it('updates the selectedLocation property of the state', async () => {
		const {findAllByText} = render(
			<Provider store={store}>
				<Home />
			</Provider>
		);

		const localizarButtons = await findAllByText('Localizar');
		expect(localizarButtons.length).toBe(3);

		// Handle Click
		fireEvent.click(localizarButtons[2]);
		const house = store.getState().houses.houses[2];
		expect(store.getState().houses.selectedLocation).toEqual([house.lat, house.long]);
	});

	it('generates and matches snapshot', async () => {
		const {findAllByText} = render(
			<Provider store={store}>
				<Home />
			</Provider>
		);
		const localizarButtons = await findAllByText('Localizar');
		expect(localizarButtons.length).toBe(3);

		const HomeComponent = renderer.create(
			<Provider store={store}>
				<Home />
			</Provider>
		);
		let tree = HomeComponent.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
