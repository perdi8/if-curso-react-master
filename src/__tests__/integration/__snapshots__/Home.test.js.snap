// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page: generates and matches snapshot 1`] = `
<div
  className="sc-jSFjdj"
>
  <div
    className="sc-gtsrHT kvaTLb"
  >
    <div
      className="sc-eCApnc hBlLbq"
    >
      <div
        className="logo"
      >
        MIPISO.com.es
      </div>
      <ul
        className="menu"
      >
        <li>
          <a
            href="/acceder"
          >
            Acceder
          </a>
        </li>
        <li>
          <a
            href="/registrarse"
          >
            Registrarse
          </a>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="sc-gtsrHT kvaTLb"
    style={
      Object {
        "backgroundColor": "#E6F2F2",
        "borderBottom": "1px solid #B5DDDD",
        "borderTop": "1px solid #B5DDDD",
      }
    }
  >
    <div
      className="sc-hKFxyN hcbMhU"
    >
      <input
        placeholder="Piso, chalet o garaje…"
        type="text"
      />
      <input
        placeholder="Madrid, Barcelona o Zaragoza…"
        type="text"
      />
      <button
        className="sc-bdnxRM hNJinF"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#46C1F7",
            "border": "none",
            "padding": "0.75rem",
          }
        }
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-search fa-w-16 "
          data-icon="search"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={
            Object {
              "color": "white",
            }
          }
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <span
          style={
            Object {
              "display": "none",
            }
          }
        >
          Buscar...
        </span>
      </button>
    </div>
  </div>
  <div
    className="sc-gtsrHT kvaTLb"
  >
    <div
      className="sc-iCoGMd gwiLdB"
    >
      <div
        className="left"
      >
        <div
          className="houses-wrapper"
        >
          <div
            className="houses"
          >
            <div
              className="sc-dlnjwi kAKtfg house"
            >
              <img
                alt=""
                src="https://api-casas.kevinmamaqi.com/image6.jpg"
              />
              <div
                className="title"
              >
                Piso en Raval
              </div>
              <div
                className="footer"
              >
                <div
                  className="price"
                >
                  1067912
                </div>
                <button
                  className="sc-bdnxRM hNJinF"
                  disabled={false}
                  onClick={[Function]}
                  style={
                    Object {
                      "backgroundColor": "#8077A3",
                    }
                  }
                >
                  <span
                    style={
                      Object {
                        "color": "white",
                        "fontSize": 14,
                      }
                    }
                  >
                    Localizar
                  </span>
                </button>
              </div>
            </div>
            <div
              className="sc-dlnjwi kAKtfg house"
            >
              <img
                alt=""
                src="https://api-casas.kevinmamaqi.com/image1.jpg"
              />
              <div
                className="title"
              >
                Piso en Poble Nou
              </div>
              <div
                className="footer"
              >
                <div
                  className="price"
                >
                  942977
                </div>
                <button
                  className="sc-bdnxRM hNJinF"
                  disabled={false}
                  onClick={[Function]}
                  style={
                    Object {
                      "backgroundColor": "#8077A3",
                    }
                  }
                >
                  <span
                    style={
                      Object {
                        "color": "white",
                        "fontSize": 14,
                      }
                    }
                  >
                    Localizar
                  </span>
                </button>
              </div>
            </div>
            <div
              className="sc-dlnjwi kAKtfg house"
            >
              <img
                alt=""
                src="https://api-casas.kevinmamaqi.com/image3.jpg"
              />
              <div
                className="title"
              >
                Chalet en Poble Sec
              </div>
              <div
                className="footer"
              >
                <div
                  className="price"
                >
                  864971
                </div>
                <button
                  className="sc-bdnxRM hNJinF"
                  disabled={false}
                  onClick={[Function]}
                  style={
                    Object {
                      "backgroundColor": "#8077A3",
                    }
                  }
                >
                  <span
                    style={
                      Object {
                        "color": "white",
                        "fontSize": 14,
                      }
                    }
                  >
                    Localizar
                  </span>
                </button>
              </div>
            </div>
          </div>
          <button
            className="sc-bdnxRM hNJinF"
            disabled={false}
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "#4480E2",
              }
            }
          >
            <span
              style={
                Object {
                  "color": "white",
                }
              }
            >
              Cargar más
            </span>
          </button>
        </div>
      </div>
      <div
        className="right"
      >
        <div
          style={
            Object {
              "width": "100%",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;
